libraries/AC_PID/AC_PID_1D.cpp

```diff
@@ -0,0 +1,248 @@
+/// @file	AC_PID_1D.cpp
+/// @brief	Generic PID algorithm
+
+#include <AP_Math/AP_Math.h>
+#include "AC_PID_1D.h"
+
+const AP_Param::GroupInfo AC_PID_1D::var_info[] = {
+    // @Param: P
+    // @DisplayName: PID Proportional Gain
+    // @Description: P Gain which produces an output value that is proportional to the current error value
+    AP_GROUPINFO("P",    0, AC_PID_1D, _kp, 0),
+
+    // @Param: I
+    // @DisplayName: PID Integral Gain
+    // @Description: I Gain which produces an output that is proportional to both the magnitude and the duration of the error
+    AP_GROUPINFO("I",    1, AC_PID_1D, _ki, 0),
+
+    // @Param: IMAX
+    // @DisplayName: PID Integral Maximum
+    // @Description: The maximum/minimum value that the I term can output
+    AP_GROUPINFO("IMAX", 2, AC_PID_1D, _kimax, 0),
+
+    // @Param: FILT
+    // @DisplayName: PID Input filter frequency in Hz
+    // @Description: Input filter frequency in Hz
+    // @Units: Hz
+    AP_GROUPINFO("FILT", 3, AC_PID_1D, _filt_E_hz, AC_PID_1D_FILT_HZ_DEFAULT),
+
+    // @Param: D
+    // @DisplayName: PID Derivative Gain
+    // @Description: D Gain which produces an output that is proportional to the rate of change of the error
+    AP_GROUPINFO("D",    4, AC_PID_1D, _kd, 0),
+
+    // @Param: D_FILT
+    // @DisplayName: D term filter frequency in Hz
+    // @Description: D term filter frequency in Hz
+    // @Units: Hz
+    AP_GROUPINFO("D_FILT", 5, AC_PID_1D, _filt_D_hz, AC_PID_1D_FILT_D_HZ_DEFAULT),
+
+    AP_GROUPEND
+};
+
+// Constructor
+AC_PID_1D::AC_PID_1D(float initial_p, float initial_i, float initial_d, float initial_ff, float initial_imax, float initial_filt_E_hz, float initial_filt_D_hz, float dt) :
+    _dt(dt)
+{
+    // load parameter values from eeprom
+    AP_Param::setup_object_defaults(this, var_info);
+
+    _kp = initial_p;
+    _ki = initial_i;
+    _kd = initial_d;
+    _kff = initial_ff;
+    _kimax = fabsf(initial_imax);
+    filt_E_hz(initial_filt_E_hz);
+    filt_D_hz(initial_filt_D_hz);
+
+    // reset input filter to first value received
+    _reset_filter = true;
+
+    memset(&_pid_info, 0, sizeof(_pid_info));
```

Comment:
Is this ever being constructed in un-zeroed memory?  i.e. on the stack?

If not, we don't need to zero.

Reply:
removed thanks.